src/backend/distributed/transaction/backend_data.c

```diff
@@ -0,0 +1,546 @@
+
+/*
+ * BackendManagementShmemSize returns the size that should be allocated
+ * on the shared memory for backend management.
+ */
+static size_t
+BackendManagementShmemSize(void)
+{
+	Size size = 0;
+
+	size = add_size(size, sizeof(BackendManagementShmemData));
+	size = add_size(size, mul_size(sizeof(BackendData), MaxBackends));
+
+	return size;
+}
+
+
+/*
+ *  InitializeBackendData is called per backend and does the
+ *  required initilizations.
+ */
+void
+InitializeBackendData(void)
+{
+	MyBackendData = &backendManagementShmemData->backends[MyProc->pgprocno];
+
+	Assert(MyBackendData);
+
+	SpinLockAcquire(&MyBackendData->mutex);
+
+	MyBackendData->databaseId = MyDatabaseId;
+	MyBackendData->processId = MyProcPid;
+	MyBackendData->transactionId.initiatorNodeIdentifier = 0;
+	MyBackendData->transactionId.transactionNumber = 0;
+	MyBackendData->transactionId.timestamp = 0;
+
+	SpinLockRelease(&MyBackendData->mutex);
+}
+
+
+/*
+ * UnSetBackendData simply acquires the mutex and resets the backend's
+ * distributed transaction data in shared memory to the initial values.
+ */
+void
+UnSetBackendData(void)
+{
+	/* backend does not exist if the extension is not created */
+	if (MyBackendData)
+	{
+		SpinLockAcquire(&MyBackendData->mutex);
+
+		MyBackendData->databaseId = 0;
+		MyBackendData->processId = 0;
+		MyBackendData->transactionId.initiatorNodeIdentifier = 0;
+		MyBackendData->transactionId.transactionNumber = 0;
+		MyBackendData->transactionId.timestamp = 0;
+
+		SpinLockRelease(&MyBackendData->mutex);
+	}
+}
+
+
+/*
+ * GetCurrentDistributedTransctionId reads the backend's distributed transaction id and
+ * returns a copy of it.
+ */
+DistributedTransactionId *
+GetCurrentDistributedTransctionId(void)
+{
+	DistributedTransactionId *currentDistributedTransactionId =
+		(DistributedTransactionId *) palloc(sizeof(DistributedTransactionId));
+
+	SpinLockAcquire(&MyBackendData->mutex);
+
+	currentDistributedTransactionId->initiatorNodeIdentifier =
+		MyBackendData->transactionId.initiatorNodeIdentifier;
+	currentDistributedTransactionId->transactionNumber =
+		MyBackendData->transactionId.transactionNumber;
+	currentDistributedTransactionId->timestamp =
+		MyBackendData->transactionId.timestamp;
+
+	SpinLockRelease(&MyBackendData->mutex);
+
+	return currentDistributedTransactionId;
+}
+
+
+/*
+ * AssignBackendData generates a new distributed transaction id and
+ * sets it for the current backend. It also sets the databaseId and
+ * processId fields.
+ *
+ * This function should only be called on BeginCoordinatedTransaction(). Any other
+ * callers is very likely to break the distributed transction management.
+ */
+void
+AssignBackendData(void)
+{
+	pg_atomic_uint64 *transactionNumberSequence =
+		&backendManagementShmemData->nextTransactionNumber;
+
+	uint64 nextTransactionNumber = pg_atomic_fetch_add_u64(transactionNumberSequence, 1);
+	TimestampTz currentTimestamp = GetCurrentTimestamp();
+	uint64 localGroupId = (uint64) GetLocalGroupId();
```

Comment:
(absolutely inconsequential, feel entirely free to discard: personally I'd do the GetLocalGroupId() first - makes the timestamp a tiny bit more accurate).